src/librustc/traits/error_reporting.rs

```diff
@@ -2342,9 +2412,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 ("`Sync`", "shared")
             };
 
-            err.clear_code();
```

Comment:
I don't have any strong feelings about undoing this change from #65345. My intent in clearing the error code was avoiding having `E0277` have multiple different primary messages:

#65345 introduced two special-cased error messages: 

- "future cannot be sent between threads safely" (special case for `Send`)
- "future cannot be shared between threads safely" (special case for `Sync`)

`E0277` otherwise has the following message:

> the trait bound `Foo: Qux` is not satisfied in `impl std::future::Future`

If we don't mind having `E0277` be the error code for all of these cases, then we should remove this line.

cc @rust-lang/wg-diagnostics

Reply:
I do find E0277 to be a too generic error that happens all over the place, and we have precedence of changing error codes for more specific cases (type errors are _not_ just E0308). We should add new error codes for these cases so that we can add explanations in the index.